### DISCUSSION TITLE
Improving mAP

### DISCUSSION THREAD

**Discussion by arrufat** [dlibml/yolo-object-detector#2] [General]
Let's try to figure out why mAP is low.

**Comment by VisionEp1** [comment]
did your yolov7 run finish?

**Reply by arrufat** [reply]
The training has now finished. It resulted in an mAP of 51.93%... not great

```
             mAP    mPr    mRc    mF1    µPr    µRc    µF1    wPr    wRc    wF1
EPOCH 93: 0.5193 0.6091 0.5113 0.5519 0.6109 0.5528 0.5804 0.6077 0.5528 0.5757
```
![image](https://user-images.githubusercontent.com/1671644/185654769-81196086-2cc8-4557-aed2-164821616a34.png)

**Reply by arrufat** [reply]
Just for fun, I ran it on the same video @davisking used for his MMOD+CNN Detector https://youtu.be/F_3uj790WOw

**Reply by VisionEp1** [reply]
thanks for the update,
did you also calculate the map on the coco eval server?

its a bit confusing that there is always the map on the testset and the map on the eval server (MS COCO evaluation server)
since we try to compare with the "official" papers I think we should also check that number

also can you share the pretrained y7 weights. so I can use them for fine tuning and further testing. might save some time.

**Reply by arrufat** [reply]
@VisionEp1, you can reproduce the results by using the `yolov7` branch.

I have attached the weights, training script, training log, loss plot, JSON for COCO evaluation and the results in [this file](https://1drv.ms/u/s!Avd3wR2lFRWkgfZhHT6UK1EuEe41Sg).

The accuracy on the test set is similar: 52.7% vs 51.93% on the validation set.
Have fun!

**Reply by pfeatherstone** [reply]
I wonder what would happen if you trained for 300 epochs and set the final min learning rate to 1e-6. The loss was still decreasing

**Reply by arrufat** [reply]
I tried something similar with the YOLOv5 architecture and I gained a couple of mAP points at best. But maybe this one behaves differently... Anyway, at some point I want to try YOLOv7 tiny to make the tests faster.

**Reply by pfeatherstone** [reply]
Nice. I reckon playing with setting different learning rates, weight decays and momenta for biases, batch norm matrices and convolutional weights would be beneficial. I  think this is possible by applying a visitor. I'll try this at some point.

**Reply by pfeatherstone** [reply]
I wonder if @glenn-jocher did any tests where all augmentation was turned off. How badly did that affect performance ?

**Reply by arrufat** [reply]
I am running a training with and without mosaic enabled for YOLOv7-tiny. The mosaic version has considerable better performance.

I made some changes to the mosaic published here, I will commit soon, together with the model if it works

It's training on my laptop, so it's a bit slow... And I stopped the non-mosaic version.

![image](https://user-images.githubusercontent.com/1671644/186157850-ec091cd2-6e23-48a3-88ad-23c22515d62f.png)

**Reply by arrufat** [reply]
By the way, YOLOv7-tiny runs at 50 fps on a 1280×720 image using 1296 MiB on a Quadro RTX 5000 Mobile.

**Comment by VisionEp1** [comment]
Hi,
i finally upgraded to 2x3090ti's and 4x2080ti's. So i can run my tests regarding batch sizes since its one of the major differences to the original repo (which works with subdivisions).

Sadly the plotting doesnt quite work for windows. But i think i might give it a shot with wsl.
The issue is even if i run the code in windows and produce the log file, the spacing seems incompatible for the plot script even if i run it in linux. So i guess i will try the whole process in linux for now to create the graphs.

Feel free to suggest any other trainings u might think will give us more insights, the hardware is ready now. Just Debugging time is limited as always ;)

**Reply by arrufat** [reply]
If you give me the training.log, I can plot it for you :)

**Reply by arrufat** [reply]
Also, I ran another training with slightly changed color augmentation (I am just using dlib's `disturb_colors(image, rnd, 1.0, 0.5)`) and for 300 epochs, which gave me:
```
             mAP    mPr    mRc    mF1    µPr    µRc    µF1    wPr    wRc    wF1
EPOCH 295: 0.5317 0.5894 0.5351 0.5572 0.6063 0.5816 0.5937 0.6020 0.5816 0.5885
```
![image](https://user-images.githubusercontent.com/1671644/195050004-8723753e-f635-47e6-9f3a-884cc398399e.png)
